src/item.cpp

```diff
@@ -3967,163 +4011,114 @@ static void eject_casings( player &p, item *reload_target, itype_id casing_type
-                    reload_target = &contents[i];
+        // In order of preference reload active gunmod, gun, spare magazine, any other auxiliary gunmod
+        std::vector<item *> opts = { active_gunmod(), this, spare_mag };
+        std::transform(contents.begin(), contents.end(), std::back_inserter(opts), [](item& mod){
+            return mod.is_auxiliary_gunmod() ? &mod : nullptr;
+        });
+
+        for( auto e : opts ) {
+            if( e != nullptr ) {
+                // @todo deprecate handling of spare magazines as a special case
+                if( e == spare_mag && ammo_type() == ammo->ammo_type() &&
+                    (!e->charges || e->get_curammo_id() == ammo->typeId()) &&
+                    e->charges < ammo_capacity() ) {
+
+                    qty = has_flag("RELOAD_ONE") ? 1 : ammo_capacity() - e->charges;
+                    target = e;
+                    break;
+                } // handle everything else (gun and auxiliary gunmods)
+                if( e->ammo_type() == ammo->ammo_type() &&
+                    (!e->ammo_remaining() || e->get_curammo_id() == ammo->typeId()) &&
+                    e->ammo_remaining() < e->ammo_capacity() ) {
+                    qty = e->ammo_capacity() - e->ammo_remaining();
+
+                    qty = e->has_flag("RELOAD_ONE") ? 1 : e->ammo_capacity() - e->ammo_remaining();
+                    target = e;
                     break;
                 }
             }
         }
-
-        if (reload_target == NULL) {
-            return false;
+    } else if( is_tool() ) {
+        qty = ammo_capacity() - ammo_remaining();
+        if( ammo_type() == ammo->ammo_type() && qty > 0 ) {
+            target = this;
         }
+    }
 
-        if (reload_target->is_gun() || reload_target->is_gunmod()) {
-            if (reload_target->is_gunmod() && reload_target->typeId() == "spare_mag") {
-                // Use gun numbers instead of the mod if it's a spare magazine
-                max_load = type->gun->clip;
-                single_load = has_flag("RELOAD_ONE");
-            } else {
-                single_load = reload_target->has_flag("RELOAD_ONE");
-                max_load = reload_target->clip_size();
-            }
+    // If we found a suitable target, try and reload it
+    if ( target ) {
+        if( has_curammo() ) {
+            eject_casings( u, target, get_curammo()->ammo->casing );
         }
-    } else if (is_tool()) {
-        const auto tool = dynamic_cast<const it_tool*>(type);
-        reload_target = this;
-        single_load = false;
-        max_load = tool->max_charges;
-    } else {
-        return false;
-    }
 
-    if (has_flag("DOUBLE_AMMO")) {
-        max_load *= 2;
-    }
+        if( target->is_gun() || target->is_gunmod() ) {
+            target->set_curammo( *ammo );
+        }
 
-    if (has_flag("ATOMIC_AMMO")) {
-        max_load *= 100;
-    }
+         if( ammo_type() == "plutonium" ) {
```

Comment:
The indentation in this function is sometime a bit strange (3 spaces per level), probably a remnant of the old code. As you're touching the code anyway, you might as well fix this (should be 4 spaces per level).